src/runtime/pprof/pprof.go

```diff
@@ -776,6 +779,107 @@ func writeGoroutineStacks(w io.Writer) error {
 	return err
 }
 
+// writeGoroutineStacksWithLabels writes individual goroutine stack traces like
+// writeGoroutineStacks, but uses the concurrent, reduced-stop time approach of
+// goroutineProfileWithLabelsConcurrent. It includes the ID, state, and labels
+// for each goroutine, although as these are captured after the world is resumed
+// they are not guaranteed to be entirely consistent as of a single point in
+// time.
+func writeGoroutineStacksWithLabels(w io.Writer) error {
+	n, ok := pprof_goroutineStacksWithLabels(nil, nil, nil)
+
+	for {
+		// Allocate slices for individual goroutine data
+		stacks := make([]profilerecord.StackRecord, n+10)
+		labels := make([]unsafe.Pointer, n+10)
+		infos := make([]profilerecord.GoroutineInfo, n+10)
+
+		n, ok = pprof_goroutineStacksWithLabels(stacks, labels, infos)
+		if ok {
+			return printGoroutineStacksWithLabels(w, stacks[:n], labels[:n], infos[:n])
+		}
+		// Profile grew; try again with larger slices
+	}
+}
+
+// printGoroutineStacksWithLabels formats goroutine records in a format similar
+// to runtime.Stack, but also includes labels as well.
+func printGoroutineStacksWithLabels(
+	w io.Writer, records []profilerecord.StackRecord, labels []unsafe.Pointer, infos []profilerecord.GoroutineInfo,
+) error {
+	for i, r := range records {
+		goroutineID := infos[i].ID
+		state := "unknown"
+		if i < len(infos) && infos[i].State != "" {
+			state = infos[i].State
+		}
+
+		// Calculate blocked time (same logic as goroutineheader in traceback.go)
+		blockedStr := ""
+		if i < len(infos) && infos[i].WaitSince != 0 {
+			// Check if this is a waiting or syscall state where blocked time applies.
+			if strings.Contains(state, "waiting") || strings.Contains(state, "syscall") {
+				waitfor := (runtime_nanotime() - infos[i].WaitSince) / 60e9 // convert to minutes
+				if waitfor >= 1 {
+					blockedStr = fmt.Sprintf(", %d minutes", waitfor)
+				}
+			}
+		}
+
+		// Get label information using the same format as debug=1.
+		var labelStr string
+		if i < len(labels) && labels[i] != nil {
+			labelMap := (*labelMap)(labels[i])
+			if labelMap != nil && len(labelMap.list) > 0 {
+				labelStr = labelMap.String()
+			}
+		}
+
+		fmt.Fprintf(w, "goroutine %d [%s%s]%s:\n", goroutineID, state, blockedStr, labelStr)
```

Comment:
one side-effect of the format change is that panicparse and friends won't accept this any more. Can you think of a way to squeeze the labels in in a way that doesn't "nominally" change the format? I can't - but thought I'd ask.

Reply:
if we wanted to use backtrace/debug=2 format we could inject them inside the `[]`; I previously had assumed that was strictly `state[, waitSince]` but looking at the 1.25 changes that add bubble info there, and the scanning states, I realized that it is expected that any number of comma separated fields can be in the `[]` so we could put labels there. But we can never get the args in a resume-the-world safe way to I think maybe trying to make debug=2 without STW is a fool's errand and we should focus on getting what we want (state, creator, waitSince) in debug=0/1, ala https://github.com/golang/go/issues/74954